### PR TITLE
Add new telemetry types

### PR DESCRIPTION
**Pull request by daesun-park**
## Description

This PR introduces a new set of Telemetry properties/interfaces, which extend the current types to include flat arrays. The following types/interfaces were introduced in `packages/utils/telemetry-utils/src/telemetryTypes.ts`.
- TelemetryEventPropertyTypeExt
- ITaggedTelemetryPropertyTypeExt
- ITelemetryPropertiesExt

The intent here is to introduce these "extended" telemetry properties, and slowly start to change our codebase to use the new `ITelemetryPropertiesExt` in place of the old "ITelemetryProperties`.

### Questions
1. Should there be any note in `Upcoming.md` regarding this?

## Does this introduce a breaking change?
No

### CODE REVIEW

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryBaseLogger, TelemetryEventCategory } from "@fluidframework/common-definitions";
+
+/**
+ * Property types that can be logged.
+ *
+ * @remarks Logging entire objects is considered extremely dangerous from a telemetry point of view because people can
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:11]: I feel like I see this comment all over the place, it's a good point but I don't think it belongs here.

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+ * easily add fields to objects that shouldn't be logged and not realize it's going to be logged.
+ * General best practice is to explicitly log the fields you care about from objects.
+ */
+export type TelemetryEventPropertyTypeExt =
+    | string
+    | number
+    | boolean
+    | undefined
+    | (string | number | boolean)[];
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:20]: @vladsud - were you suggesting we support `null` in here too? Kind of hard to imagine needing it (a list of optional numbers or something...?), but in the interest of being a flexible API seems worth including. I know we shouldn't do `undefined` since `JSON.stringify([undefined]) === "[null]"`.

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+ /**
+  * A property to be logged to telemetry containing both the value and a tag. Tags are generic strings that can be used
+  * to mark pieces of information that should be organized or handled differently by loggers in various first or third
+  * party scenarios. For example, tags are used to mark PII that should not be stored in logs.
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:25]: The term "PII" is discouraged. This is probably copied/pasted, but can you update it?  Say "personal information" instead, it's generic enough.

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    eventName: string;
+}
+
+/**
+ * Interface to output telemetry events.
+ * Implemented by hosting app / loader
+ */
+ export interface ITelemetryBaseLoggerExt {
+    /**
+     * An optional boolean which indicates to the user of this interface that tags (i.e. `ITaggedTelemetryPropertyType`
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:56]: nit
```suggestion
     * An optional boolean which indicates to the user of this interface that tags (i.e. `ITaggedTelemetryPropertyTypeExt`
```

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+ export interface ITelemetryBaseEventExt extends ITelemetryPropertiesExt {
+    category: string;
+    eventName: string;
+}
+
+/**
+ * Interface to output telemetry events.
+ * Implemented by hosting app / loader
+ */
+ export interface ITelemetryBaseLoggerExt {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:54]: I don't think we want this, is it needed somewhere?  "Base" terminology should be reserved for the old types used across public interface boundaries.

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    /**
+     * Return all properties from this object that should be logged to telemetry
+     */
+    getTelemetryProperties(): ITelemetryPropertiesExt;
+}
+
+/**
+ * ITelemetryLogger interface contains various helper telemetry methods,
+ * encoding in one place schemas for various types of Fluid telemetry events.
+ * Creates sub-logger that appends properties to all events
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:103]: I don't think this line belongs.

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    [index: string]: TelemetryEventPropertyTypeExt | ITaggedTelemetryPropertyTypeExt;
+}
+
+/**
+ * Base interface for logging telemetry statements.
+ * Can contain any number of properties that get serialized as json payload.
+ * @param category - category of the event, like "error", "performance", "generic", etc.
+ * @param eventName - name of the event.
+ */
+ export interface ITelemetryBaseEventExt extends ITelemetryPropertiesExt {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:45]: Let's call this `ITelemetryEventExt`, I want to reserve the "Base" terminology for the old types used across public interface boundaries.

Comment (markfields) [reply packages/utils/telemetry-utils/src/telemetryTypes.ts:45]: Or maybe the "Ext" version is not even needed

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    sendErrorEvent(event: ITelemetryErrorEventExt, error?: any): void;
+
+    /**
+     * Send performance telemetry event
+     * @param event - Event to send
+     */
+    sendPerformanceEvent(event: ITelemetryPerformanceEventExt, error?: any): void;
+}
+
+export type TelemetryEventTypes =
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:133]: I think you should be able to use `ITelemetryGenericEventExt` instead of defining this new type.

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+ * encoding in one place schemas for various types of Fluid telemetry events.
+ * Creates sub-logger that appends properties to all events
+ */
+export interface ITelemetryLoggerExt extends ITelemetryBaseLogger {
+    /**
+     * Actual implementation that sends telemetry event
+     * Implemented by derived classes
+     * @param event - Telemetry event to send over
+     */
+    send(event: ITelemetryBaseEventExt): void;
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:111]:
```suggestion
    send(event: ITelemetryBaseEvent): void;
```

Not sure why this is here on the old def's anyway, it's redundant with `extends ITelemetryBaseLogger`.  I would remove it.  @vladsud any reason this interface should explicitly mention `send` when it's already on the base interface?

File: packages/utils/telemetry-utils/src/logger.ts [hunk 10]
```diff
@@ -139,9 +147,9 @@ export abstract class TelemetryLogger implements ITelemetryLogger {
      * @param error - optional error object to log
      */
      protected sendTelemetryEventCore(
-        event: ITelemetryGenericEvent & { category: TelemetryEventCategory; },
+        event: ITelemetryGenericEventExt & { category: TelemetryEventCategory; },
         error?: any) {
-        const newEvent = { ...event };
+        const newEvent = convertToBaseEvent(event) as ITelemetryBaseEventExt;
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:152]: I'm wondering if I'm missing something, but this should be fine to just be the old `ITelemetryBaseEvent`
```suggestion
        const newEvent = convertToBaseEvent(event);
```

File: packages/utils/telemetry-utils/src/logger.ts [hunk 11]
```diff
@@ -185,11 +193,9 @@ export abstract class TelemetryLogger implements ITelemetryLogger {
         this.sendTelemetryEventCore(perfEvent, error);
     }
 
-    protected prepareEvent(event: ITelemetryBaseEvent): ITelemetryBaseEvent {
+    protected prepareEvent(event: ITelemetryBaseEventExt): ITelemetryBaseEventExt {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:196]: Side note - Not sure why this function isn't used in this base class... Not related to this PR but maybe a small fix to discuss later.

File: packages/utils/telemetry-utils/src/logger.ts [hunk 12]
```diff
@@ -185,11 +193,9 @@ export abstract class TelemetryLogger implements ITelemetryLogger {
...
         const includeErrorProps = event.category === "error" || event.error !== undefined;
-        const newEvent: ITelemetryBaseEvent = {
-            ...event,
-        };
+        const newEvent: ITelemetryBaseEventExt = convertToBaseEvent(event) as ITelemetryBaseEventExt;
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:198]: Shouldn't need to do this, or really change this function at all.  It's dealing in the base event type and should stay that way (by this time extended events have already been converted)

File: packages/utils/telemetry-utils/src/logger.ts [hunk 13]
```diff
@@ -587,3 +593,43 @@ export class TelemetryNullLogger implements ITelemetryLogger {
...
     public sendPerformanceEvent(event: ITelemetryPerformanceEvent, error?: any): void {}
 }
+
+/**
+ * Take in a event object, stringify any fields that are non-primitives, and return the new event object.
+ * @param event - Event with fields you want to stringify.
+ */
+ function convertToBaseEvent(event: TelemetryEventTypes): TelemetryEventTypes {
+    const newEvent = {
+        ...event
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:603]: Not sure you want to do this.  Since `newEvent[key]` is not overwritten if `convertToBasePropertyType` returns `null`, then say there is some arbitrary object on here -- it will stay around.  This is of course how this code used to work, but if we're introducing this conversion layer I think it should be pretty strict.

Best way to approach these corner cases is to try to enumerate them in unit tests. Then you can see for sure the current behavior and be deliberate about the final behavior (this is why they're called `.spec.ts` - they specify the behavior of the code under test)

Comment (markfields) [reply packages/utils/telemetry-utils/src/logger.ts:603]: I think we can follow the precedent in the function `[filterValidTelemetryProps](https://github.com/markfields/FluidFramework/blob/89bb5e66a85f67a0e40583676feb922a0d0bdc5e/packages/utils/telemetry-utils/src/errorLogging.ts#L283-L285)`, and log a console error (to catch someone's eye while debugging) and the set a stock value like `"INVALID PROPERTY (arbitrary object)"` (use a different value than that error logging code).

Probably put that code in `convertToBasePropertyType` actually since it's the code that knows why it's invalid.

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+ * Maps to category = "performance"
+ */
+export interface ITelemetryPerformanceEventExt extends ITelemetryGenericEventExt {
+    duration?: number; // Duration of event (optional)
+}
+
+/**
+ * An error object that supports exporting its properties to be logged to telemetry
+ */
+export interface ILoggingErrorExt extends Error {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:93]: I think we decided we wouldn't change `getTelemetryProperties` stuff, so this interface isn't necessary.  We might want to update `setTelemetryProperties` to accept the Ext properties though.

Comment (markfields) [reply packages/utils/telemetry-utils/src/telemetryTypes.ts:93]: But we _should_ update `addTelemetryProperties` to take in `ITelemetryPropertiesExt`, and then [`getValidTelemetryProps`](https://github.com/markfields/FluidFramework/blob/89bb5e66a85f67a0e40583676feb922a0d0bdc5e/packages/utils/telemetry-utils/src/errorLogging.ts#L303) will convert just like your functions here (maybe you can even share code at some point)

**Review by markfields — COMMENTED** [review]
Getting really close!  Biggest feedback is that we shouldn't need `ITelemetryBaseLoggerExt` or `ITelemetryBaseEventExt`.

Before merging this, I'd love to see a draft PR branched off this branch that starts to migrate call sites over to the new interfaces to confirm this will pan out as expected.

File: packages/utils/telemetry-utils/src/telemetryTypes.ts (hunk 1 above)
Comment (daesun-park) [reply packages/utils/telemetry-utils/src/telemetryTypes.ts]: removed comment

File: packages/utils/telemetry-utils/src/telemetryTypes.ts (hunk 5 above)
Comment (daesun-park) [reply packages/utils/telemetry-utils/src/telemetryTypes.ts:50]: removed interface

File: packages/utils/telemetry-utils/src/telemetryTypes.ts (hunk 6 above)
Comment (daesun-park) [reply packages/utils/telemetry-utils/src/telemetryTypes.ts]: removed comment

File: packages/utils/telemetry-utils/src/telemetryTypes.ts (hunk 8 above)
Comment (daesun-park) [reply packages/utils/telemetry-utils/src/telemetryTypes.ts]: removed interface, replaced with `ITelemetryGenericEventExt`

File: packages/utils/telemetry-utils/src/telemetryTypes.ts (hunk 9 above)
Comment (daesun-park) [reply packages/utils/telemetry-utils/src/telemetryTypes.ts]: removed method

File: packages/utils/telemetry-utils/src/logger.ts (hunk 10 above)
Comment (daesun-park) [reply packages/utils/telemetry-utils/src/logger.ts]: committed suggested change

File: packages/utils/telemetry-utils/src/logger.ts (hunk 12 above)
Comment (daesun-park) [reply packages/utils/telemetry-utils/src/logger.ts]: I reverted back to not calling `convertToBaseEvent(event)`, but would this be okay for new logger classes that extend `TelemetryLogger` and implement it's own send shown in the example below?

```
class TestTelemetryLogger extends TelemetryLogger {
    public events: ITelemetryBaseEventExt[] = [];
    public send(event: ITelemetryBaseEventExt): void {
        this.events.push(this.prepareEvent(event));
    }
}
```

If the extended event is passed through directly into the `prepareEvent()` function (as shown above), it will not be converted yet and lead to an error.

File: packages/utils/telemetry-utils/src/logger.ts (hunk 13 above)
Comment (daesun-park) [reply packages/utils/telemetry-utils/src/logger.ts:603]: updated `convertToBasePropertyType` to return `"INVALID PROPERTY (arbitrary object)"` at the end of the default case, but did you mean that we should return this in place of `null` (shown below)?

```
            if (!Array.isArray(x)) {
                return null;
            }
```
change to 
```
            if (!Array.isArray(x)) {
                console.error(`UnSupported Format of Logging Property for key ${key}:`, x);
                return "INVALID PROPERTY (arbitrary object)";
            }
```

File: packages/utils/telemetry-utils/src/telemetryTypes.ts (hunk 14 above)
Comment (daesun-park) [reply packages/utils/telemetry-utils/src/telemetryTypes.ts]: removed interface, updated `getValidTelemetryProps`

File: packages/utils/telemetry-utils/src/logger.ts (hunk 13 above)
Comment (markfields) [reply packages/utils/telemetry-utils/src/logger.ts:603]: Yeah. And remove the null check on line 607 below.  I'm adding some comments in `convertToBasePropertyType` too, take a look.

File: packages/utils/telemetry-utils/src/logger.ts [hunk 15]
```diff
@@ -587,3 +593,45 @@ export class TelemetryNullLogger implements ITelemetryLogger {
...
+    }
+    return newEvent as ITelemetryBaseEvent;
+}
+
+/**
+ * Takes in parameter, if parameter is of primitive type, return the original value.
+ * If parameter is an array, stringify then return the result.
+ * @param x - parameter passed to validate/filter
+ */
+function convertToBasePropertyType(x: any, key: string): TelemetryEventPropertyType | null {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:619]:
```suggestion
function convertToBasePropertyType(x: any, key: string): TelemetryEventPropertyType {
```

File: packages/utils/telemetry-utils/src/logger.ts [hunk 16]
```diff
@@ -587,3 +593,45 @@ export class TelemetryNullLogger implements ITelemetryLogger {
...
+    switch (typeof x) {
+        case "string":
+        case "number":
+        case "boolean":
+        case "undefined":
+            return x;
+        default:
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:626]: Add case for "object" and do all this in there.  Then in the default case return `INVALID PROPERTY (typed as ${typeof x})` or something

File: packages/utils/telemetry-utils/src/logger.ts [hunk 17]
```diff
@@ -587,3 +593,45 @@ export class TelemetryNullLogger implements ITelemetryLogger {
...
+            if (!Array.isArray(x)) {
+                return null;
+            }
+            if (x.every((val) => typeof val === "boolean" || typeof val === "string" || typeof val === "number")) {
+                return JSON.stringify(x);
+            }
+            // We don't support logging arbitrary objects
+            console.error(`UnSupported Format of Logging Property for key ${key}:`, x);
+            return "INVALID PROPERTY (arbitrary object)";
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:635]:
```suggestion
            return "INVALID PROPERTY (array with unsupported values)";
```

File: packages/utils/telemetry-utils/src/telemetryTypes.ts (hunk 5 above)
Comment (markfields) [reply packages/utils/telemetry-utils/src/telemetryTypes.ts:50]: But it's still here?

File: packages/utils/telemetry-utils/src/logger.ts [hunk 18]
```diff
@@ -12,8 +12,8 @@ import {
     ITelemetryPerformanceEvent,
     ITelemetryProperties,
     TelemetryEventPropertyType,
-    ITaggedTelemetryPropertyType,
     TelemetryEventCategory,
+    ITaggedTelemetryPropertyType,
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:16]: super nit: Nice to revert unnecessary changes to reduce code churn

File: packages/utils/telemetry-utils/src/logger.ts [hunk 19]
```diff
@@ -587,3 +595,45 @@ export class TelemetryNullLogger implements ITelemetryLogger {
     public sendErrorEvent(event: ITelemetryErrorEvent, error?: any): void {}
     public sendPerformanceEvent(event: ITelemetryPerformanceEvent, error?: any): void {}
 }
+
+/**
+ * Take in a event object, stringify any fields that are non-primitives, and return the new event object.
+ * @param event - Event with fields you want to stringify.
+ */
+ function convertToBaseEvent({ category, eventName, ...props }: ITelemetryEventExt): ITelemetryBaseEvent {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:603]:
```suggestion
function convertToBaseEvent({ category, eventName, ...props }: ITelemetryEventExt): ITelemetryBaseEvent {
```

File: packages/utils/telemetry-utils/src/errorLogging.ts [hunk 20]
```diff
@@ -405,7 +406,7 @@ export class LoggingError extends Error implements ILoggingError, Omit<IFluidErr
     /**
      * Add additional properties to be logged
      */
-    public addTelemetryProperties(props: ITelemetryProperties) {
+    public addTelemetryProperties(props: ITelemetryPropertiesExt) {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/errorLogging.ts:409]: Let's not do this in this PR. It will not be too much work, but there are a few additional considerations so better to scope it out.  I opened https://dev.azure.com/fluidframework/internal/_workitems/edit/2802 for this.

File: packages/utils/telemetry-utils/src/debugLogger.ts [hunk 21]
```diff
@@ -81,7 +81,7 @@ export class DebugLogger extends TelemetryLogger {
      * @param event - the event to send
      */
     public send(event: ITelemetryBaseEvent): void {
-        const newEvent: ITelemetryProperties = this.prepareEvent(event);
+        const newEvent: ITelemetryPropertiesExt = this.prepareEvent(event);
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/debugLogger.ts:84]: Not necessary - `event` is already a base event.

File: packages/utils/telemetry-utils/src/logger.ts [hunk 22]
```diff
@@ -587,3 +595,45 @@ export class TelemetryNullLogger implements ITelemetryLogger {
     public sendErrorEvent(event: ITelemetryErrorEvent, error?: any): void {}
     public sendPerformanceEvent(event: ITelemetryPerformanceEvent, error?: any): void {}
 }
+
+/**
+ * Take in a event object, stringify any fields that are non-primitives, and return the new event object.
+ * @param event - Event with fields you want to stringify.
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:601]: Can you please update this comment? Just something high level, doesn't need to mention stringify (and it's just not accurate at the moment)

File: packages/utils/telemetry-utils/src/logger.ts [hunk 23]
```diff
@@ -587,3 +595,45 @@ export class TelemetryNullLogger implements ITelemetryLogger {
...
+    const newEvent: ITelemetryBaseEvent = { category, eventName };
+    for (const key of Object.keys(props)) {
+        newEvent[key] = convertToBasePropertyType(props[key]);
+    }
+    return newEvent;
+}
+
+/**
+ * Takes in parameter, if parameter is of primitive type, return the original value.
+ * If parameter is an array, stringify then return the result.
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:613]: Same here - update this comment please

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * Performance telemetry event.
+ * Maps to category = "performance"
+ */
+export interface ITelemetryPerformanceEventExt extends ITelemetryGenericEventExt {
+    duration?: number; // Duration of event (optional)
+}
+
+export interface ITelemetryLoggerExt extends ITelemetryBaseLogger {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:71]: Comment please.  Maybe mention that this is for use internally within FF code, and ITelemetryBaseLogger should be used when loggers are passed between layers.

File: packages/utils/telemetry-utils/src/logger.ts [hunk 25]
```diff
@@ -85,7 +93,7 @@ export abstract class TelemetryLogger implements ITelemetryLogger {
      * @param error - Error to extract info from
      * @param fetchStack - Whether to fetch the current callstack if error.stack is undefined
      */
-    public static prepareErrorObject(event: ITelemetryBaseEvent, error: any, fetchStack: boolean) {
+    public static prepareErrorObject(event: ITelemetryEventExt, error: any, fetchStack: boolean) {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:96]: Revert this

File: packages/utils/telemetry-utils/src/logger.ts [hunk 26]
```diff
@@ -587,3 +595,52 @@ export class TelemetryNullLogger implements ITelemetryLogger {
...
+        case "boolean":
+        case "undefined":
+            return x;
+        case "object":
+            if (Array.isArray(x)) {
+                // We know that the array type is
+                return JSON.stringify(x);
+            }
+            return {
+                value: JSON.stringify(
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:636]: Try this test case:  `convertToBasePropertyType({value: true, tag: foo })`.  I don't think you want the `JSON.stringify`.

**Review by markfields — APPROVED** [review]
There's one comment to address (a correctness problem I think), and tests to write, but once those are addressed this looks good to me!

https://github.com/microsoft/FluidFramework/pull/12863/files#r1042817452

File: packages/utils/telemetry-utils/src/logger.ts (hunk 12 above)
Comment (markfields) [reply packages/utils/telemetry-utils/src/logger.ts]: Not sure I follow so lmk if I'm missing it.  But no one should ever override `send` and change its type - we just add the specialized `sendFoo` functions that take the extended stuff.  But `send` itself needs to stay basic.

File: packages/utils/telemetry-utils/src/logger.ts [hunk 27]
```diff
@@ -587,3 +595,52 @@ export class TelemetryNullLogger implements ITelemetryLogger {
...
+                    convertToBasePropertyType(x.value)
+                ),
+                tag: x.tag,
+            };
+        default:
+            // should never reach this case.
+            console.error("");
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:643]: Reminder to fill this in

File: packages/utils/telemetry-utils/src/logger.ts [hunk 28]
```diff
@@ -587,3 +595,52 @@ export class TelemetryNullLogger implements ITelemetryLogger {
...
+function convertToBasePropertyType(
+    x: TelemetryEventPropertyTypeExt | ITaggedTelemetryPropertyTypeExt,
+): TelemetryEventPropertyType | ITaggedTelemetryPropertyType {
+    switch (typeof x) {
+        case "string":
+        case "number":
+        case "boolean":
+        case "undefined":
+            return x;
+        case "object":
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:630]: Reminder to write tests that pass in _invalid_ objects to make sure the result is acceptable.  People "shouldn't" circumvent TypeScript type system but if they do, it would be good to have a clear illustration of what will happen.
